File: problems_list/剑指OfferII101.分割等和子集/剑指OfferII101-分割等和子集.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function(nums) {
    const n=nums.length;
    if(n<2){
        return false;
    }

    let sum=0, maxNum=0;
    for(const num of nums){
        sum+=num;
        maxNum=maxNum>num?maxNum:num;
    }

    if(sum&1){
        return false;
    }

    const target=Math.floor(sum/2);
    const dp=new Array(n).fill(0).map(()=>new Array(target+1, false));

    for(let i=0; i<n; i++){
        dp[i][0]=true;
    }
    dp[0][nums[0]]=true;
    for(let i=1; i<n; i++){
        for(let j=1; j<=target; j++){
            const num=nums[i];
            if(j>=num){
                dp[i][j]=dp[i-1][j-num]|dp[i-1][j];
            }else{
                dp[i][j]=dp[i-1][j];
            }
        }
    }
    return dp[n-1][target];
};

File: problems_list/剑指OfferII101.分割等和子集/剑指OfferII101-分割等和子集.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int n=nums.size();
        if(n<2) return false;
        
        int ssum=accumulate(nums.begin(), nums.end(), 0);
        int maxNum=*max_element(nums.begin(), nums.end());
        
        if(ssum&1) return false;
        if(maxNum>ssum/2) return false;

        int tar=ssum/2;
        vector<vector<int>> f(n, vector<int>(tar+1, 0));
        for(int i=0; i<n; i++) f[i][0]=1;
        f[0][nums[0]]=1;

        for(int i=1; i<n; i++){
            for(int j=1; j<=tar; j++){
                f[i][j]=f[i-1][j];
                if(j>=nums[i]) f[i][j]|=f[i-1][j-nums[i]];
            }
        }
        return f[n-1][tar];
    }
};